Modbus support + remote only mode

M3m3M4n opened the pull request:
I added support for modbus TCP and a remote only mode for fuzzing black box device. Modbus request / response is split to sections as per modbus specification description.

Since modbus is a binary protocol, there is no clear distinction between packets. Keeping current implementation means we'll have to trust size field in request / response. 

Changes also have to be made to the sending steps to merge message before sending since some modbus implementation (e.g. pymodbus) fail to process if we send them sequentially.

thuanpv commented:
Hi @M3m3M4n. Thank you very much for your contribution! Adding support for Modbus is great. I have quickly looked at your changes and it LGTM. To ensure that the new protocol works and your changes do not affect the existing features/supported protocols, we need to conduct some tests. It would be great if you could add a tutorial for this newly supported protocol, following the structure of other tutorials at https://github.com/aflnet/aflnet/tree/master/tutorials. Please feel free to let us know if you need any help.

M3m3M4n commented:
Hi @thuanpv, I've added example and scripts for testing modbus tcp and remote only mode.

thuanpv commented:
Thanks @M3m3M4n. I have done a quick check and it seems that the current README is somewhat incomplete so that I got some error message as below when I run the command: "afl-fuzz -d -i reqdump -o out -N tcp://127.0.0.1/5020 -P MODBUSTCP -D 10000 -q 3 -s 3 -E -R -r -W 5 -e remote_only_runserver.sh"

```
[-] Oops, the program crashed with one of the test cases provided. There are
    several possible explanations:

    - The test case causes known crashes under normal working conditions. If
      so, please remove it. The fuzzer should be seeded with interesting
      inputs - but not ones that cause an outright crash.

    - The current memory limit (50.0 MB) is too low for this program, causing
      it to die due to OOM when parsing valid files. To fix this, try
      bumping it up with the -m setting in the command line. If in doubt,
      try something along the lines of:

      ( ulimit -Sv $[49 << 10]; /path/to/binary [...] <testcase )

      Tip: you can use http://jwilk.net/software/recidivm to quickly
      estimate the required amount of virtual memory for the binary. Also,
      if you are using ASAN, see docs/notes_for_asan.txt.

    - Least likely, there is a horrible bug in the fuzzer. If other options
      fail, poke <lcamtuf@coredump.cx> for troubleshooting tips.
```
It seems to me that the instruction to setup of the server under test is not clear. I did try to run the server alone but I got another error message (as shown below). It seems that I miss components like pymodbus.
```
Traceback (most recent call last):
  File "server.py", line 5, in <module>
    from pymodbus.datastore import ModbusSequentialDataBlock
ModuleNotFoundError: No module named 'pymodbus'
```
Would you mind update your README to make it more complete? Thanks

M3m3M4n commented:
Yes I forgot pymodbus in the instruction. Added.

thuanpv commented:
Thanks @M3m3M4n. After installing the missing packages, I think the server is working now (I use the command "server.py lo 5020"). However, fuzzing cannot proceed because the given seed input causes the server to crash. Do you know why? I assume the sample input inside the reqdump folder is a benign input, am I right?

```
[*] Attempting dry run with 'id:000000,orig:dump.bin'...

[-] Oops, the program crashed with one of the test cases provided. There are
    several possible explanations:

    - The test case causes known crashes under normal working conditions. If
      so, please remove it. The fuzzer should be seeded with interesting
      inputs - but not ones that cause an outright crash.

    - The current memory limit (50.0 MB) is too low for this program, causing
      it to die due to OOM when parsing valid files. To fix this, try
      bumping it up with the -m setting in the command line. If in doubt,
      try something along the lines of:

      ( ulimit -Sv $[49 << 10]; /path/to/binary [...] <testcase )

      Tip: you can use http://jwilk.net/software/recidivm to quickly
      estimate the required amount of virtual memory for the binary. Also,
      if you are using ASAN, see docs/notes_for_asan.txt.

    - Least likely, there is a horrible bug in the fuzzer. If other options
      fail, poke <lcamtuf@coredump.cx> for troubleshooting tips.

[-] PROGRAM ABORT : Test case 'id:000000,orig:dump.bin' results in a crash
         Location : perform_dry_run(), afl-fuzz.c:3800
```

M3m3M4n commented:
Hi @thuanpv, I assume you are testing remote only mode? Fuzzer will report crash iff it encounters socket error during sending / receiving. The provided test case is benign and won't crash the server. I suspect something else. You can test the input by piping input to the server through netcat.

In remote only mode, user should use a monitoring script which monitor and restart the server if its not running, this is called every fuzz cycle to check. So you don't need to start the server by hand. A setup script is included in the tutorial section for remote only mode, just need root for mounting ramdisk, everything else run as user. Can you test with it and see if fuzzer still crash?

thuanpv commented:
```
In remote only mode, user should use a monitoring script which monitor and restart the server if its not running, this is called every fuzz cycle to check. So you don't need to start the server by hand.
```

Just wanted to clarify that I ran the server alone only for testing purpose. For fuzzing, I used the given remote_only_runserver.sh script and ran this command "afl-fuzz -d -i reqdump -o out -N tcp://127.0.0.1/5020 -P MODBUSTCP -D 10000 -q 3 -s 3 -E -R -r -W 5 -e ./remote_only_runserver.sh" as instructed in the tutorial README. I think what I missed or did wrongly is related this part of your reply "A setup script is included in the tutorial section for remote only mode, just need root for mounting ramdisk, everything else run as user." I haven't setup any specific ramdisk and I ran the fuzzing command with normal user, not with root. Could you please elaborate a bit more on that? Moreover, what are the differences between the two tutorials named "modbustcp" and "remote_only"?

M3m3M4n commented:
I think I should explain what I added in details.

> what are the differences between the two tutorials named "modbustcp" and "remote_only"?

Remote only mode works like afl dumb_mode. Means blackbox fuzzing, no afl instrumentation, only now sending payload through network layer instead of feeding them through stdin or file ... Crash detection is done via return value of send and recv (see below). Some guidance is provided by your algorithm. Should work with all protocol.

Modbustcp is the protocol I added.

some details of remote only mode:
- In run_target() I add a call to system(reset_script). reset_script is the monitoring script, which run every time run_target is called. This reset_script should check whether the fuzzing target is alive or not, and restart it somehow if it died (or not yet started - same thing). Not the best way of doing so, but will work for the time being. In my example reset_script is a bash script.
- Other than this, the fuzzer detects crash by evaluating return value of send and recv. I modified those to make them more...uniform, you can check in the code.

> "I haven't setup any specific ramdisk and I ran the fuzzing command with normal user, not with root. Could you please elaborate a bit more on that?"
 
The script remote_only_fuzz.sh should setup the environment for you, including ramdisk etc... But it need root to mount ramdisk.

>  For fuzzing, I used the given remote_only_runserver.sh script and ran this command "afl-fuzz -d -i reqdump -o out -N tcp://127.0.0.1/5020 -P MODBUSTCP -D 10000 -q 3 -s 3 -E -R -r -W 5 -e ./remote_only_runserver.sh"

Yes that is the command in the readme file, but you should prepare the environment as in the remote_only_fuzz.sh (or run it) as the included script has hardcoded path. I think this is why your fuzzer crash. The command in the readme is included as the example of how you should run it once you have your env all set up, thus this line in the readme:
>- copy the content of this directory to aflnet parent directory then run remote_only_fuzz.sh as root.

I think thats all.

16201535wangshenyu commented:
Hi @M3m3M4n, There is something wrong when I try to run the tutorial named  modbustcp. 
details as follows:
![image](https://user-images.githubusercontent.com/51477879/162357580-1d4baf06-9bc4-4885-b8e2-6912467f8a2e.png)
The raw data:
dump.bin and 
![52066-1502](https://user-images.githubusercontent.com/51477879/162358187-69df208b-e264-4aa3-9fa1-b3d382a8c1df.png)
please rename the last file to "52066-1502.raw"

M3m3M4n commented:
@16201535wangshenyu I don't encounter any error running with your data, what is the problem exactly?

Alicheche commented:
Hi@M3m3M4n， When I try to run the tutorial named modbustcp, I encountered the same error situation. I use the document dump.bin you provided. I repeated the experiment several times, and each time I quit within a minute.The program exits and prompts for a segment error.details as follows:
![图片1](https://user-images.githubusercontent.com/57059677/165015960-65117189-9cba-44be-83a8-e0757fb43516.png)
Previous attempts to test libmodbus using AFL did not encounter such errors.

M3m3M4n commented:
@Alicheche Its really hard to tell what is going on with only a segfault message, can you provide more information? like whats your OS version? Did the error occurs with the provided instrumented_fuzz.sh? if not whats your setup? can you send me your afl binary and core dump?

16201535wangshenyu commented:
Hi @M3m3M4n [M3m3M4n](https://github.com/M3m3M4n)
I have solved this problem. This is  because  I rewrite the function extract_requests_modbustcp.
``

16201535wangshenyu commented:
I have some questions to discuss, but my English is not very good. Are you Chinese, [M3m3M4n](https://github.com/M3m3M4n) [Alicheche](https://github.com/Alicheche)

M3m3M4n commented:
I don't speak chinese, but google translate should be enough to get the point across

Alicheche commented:
Hi, @M3m3M4n 
The above problem occurs after running the following command:
`afl-fuzz -d -i reqdump -o out -N tcp://127.0.0.1/1502 -P MODBUSTCP -D 10000 -q 3 -s 3 -E -R  -K random-test-server`
My OS version is ubuntu 16.04. The afl(AFLNET) binary and core dump are here: 
[https://github.com/Alicheche/aflnet-modbustcp-question](url)
Some of the error messages in the core file are as follows:
![2](https://user-images.githubusercontent.com/57059677/165421469-e3eab9da-9022-45c8-8f37-d53f60684a6d.png)
The error did not occur when  run the provided instrumented_fuzz.sh. But the program stops after a short time：
![3](https://user-images.githubusercontent.com/57059677/165422455-972a5f57-35a3-4796-8d4c-9097fd6dc89c.png)
I am new to this field, looking forward to your advice

Alicheche commented:
@16201535wangshenyu Yes, I am Chinese, very happy to communicate with you. Maybe we can add WeChat? My profile page has my contact details。

thuanpv commented:
Hi @M3m3M4n FYI, I am working towards some critical deadlines for papers, grants etc in ~2 more weeks. Will get back straight to this, do some more tests and merge your PR once I am done with those deadlines. Your PR adds an important feature to AFLNet so it is in my priority list.

wyunc commented:
> Hi, @M3m3M4n The above problem occurs after running the following command: `afl-fuzz -d -i reqdump -o out -N tcp://127.0.0.1/1502 -P MODBUSTCP -D 10000 -q 3 -s 3 -E -R -K random-test-server` My OS version is ubuntu 16.04. The afl(AFLNET) binary and core dump are here: [https://github.com/Alicheche/aflnet-modbustcp-question](url) Some of the error messages in the core file are as follows: ![2](https://user-images.githubusercontent.com/57059677/165421469-e3eab9da-9022-45c8-8f37-d53f60684a6d.png) The error did not occur when run the provided instrumented_fuzz.sh. But the program stops after a short time： ![3](https://user-images.githubusercontent.com/57059677/165422455-972a5f57-35a3-4796-8d4c-9097fd6dc89c.png) I am new to this field, looking forward to your advice

I encountered the same problem, after running afl-fuzz for some time (tens of seconds to tens of minutes) the execution was interrupted and a "segmentation fault" appeared underneath. How did you solve it?

wyunc commented:
@M3m3M4n Hello and thank you for aflnet-modbus. But I had some problems running it.

I configured aflnet-modbus_remote according to your readme. There were some problems when fuzzy testing libmodbus.
After I downloaded libmodbus from github, I used./autogen.sh to generate configure, then CC=afl-gcc, CXX=afl-g++./configure --enable-static --disable-shared, and also changed CFLAGS/CXXFLAGS/LdFLAGS in makefile.
I used the aflnet-modbus_remote provided by you to test /libmodbus/tests/unit-test-server, and the command is: /afl-fuzz -d -i in -o out -n tcp://127.0.0.1/1502 -p MODBUSTCP -d 10000 -q 3-s 3-E-R-K.. /libmodbus/tests/unit-test-server -m none
However, after running for some time (a few seconds to ten minutes) the fuzzing will break down, prompting a "segmentation fault", but when I test the random-test-server file you provided, this problem does not arise.
Is there any difference between compilation of your random-test-server and libmodbus? How can I solve this problem? Very confused

16201535wangshenyu commented:
@M3m3M4n @thuanpv . in aflnet.c , Hello and thank you for aflnet-modbus. I have some questions about it, every request message is saved in struct region_t . In extract_requests_modbustcp func,  I find that every filed (such tracid\protocol id \ length..)of modbus request message is saved in region_t struct , why not a whole message is saved in a region_t struct.

M3m3M4n commented:
@16201535wangshenyu It has been a while so I am not remember things correctly, but every region_t struct will go to the mutation algorithm. Since the data part might varies (both in length and content), I think its best to separate it into 1 single region. For the rest of the mbap header, I don't think its best to divide them further because 1, its a binary protocol, 2 the length is fixed.
